### PR TITLE
Treat so-called missing nodes as error nodes.

### PR DESCRIPTION
**Pull request by mjambon**
This causes ocaml-tree-sitter parsers to report so-called missing nodes inserted by tree-sitter. These nodes are usually wrong but we leave them in place because it's the easiest way to get a well-formed parse tree.

Closes https://github.com/returntocorp/ocaml-tree-sitter-core/issues/8

## Earlier discussion

Unlike parsing errors resulting in error nodes, this will result in a total failure to parse the file, most of the time (there's a comment in the code that briefly explains this). I'm still pondering the implications of this. It would be really nice if we could excise the minimum subtree containing the missing node rather than most of it.

The test grammar is nice because it's of the form repeat(seq()) and the presence of a missing node in the seq removes the whole seq without disrupting the repeat. However, if we had a seq(foo, seq()), removing the broken node matching the inner seq as we do still breaks the outer seq.

Update: yes, I think we should try hard to not cause full parsing failures in case of missing nodes. See 2 solutions below.

### Alternative 1

This consists in removing the missing (= made-up) nodes added by tree-sitter. We can only remove optional elements from the typed OCaml tree (typed CST). These are list and option elements. The implementation would consist in raising a `Missing_node` exception where such a node is encountered and catching `Missing_node` exceptions when attempting to process an optional element.

The result would be a tree that missing a possibly large subtree like for error nodes.

### Alternative 2

Keep the missing nodes in the tree and use them as regular nodes to form the typed CST. Additionally, report missing nodes as errors like error nodes.

This would make minimal damage to the typed CST and is easy to implement. I think we should do this.

### Security

- [x] Change has no security implications (otherwise, ping the security team)

### CODE REVIEW

**Comment by aryx** [comment]
This closes https://github.com/returntocorp/ocaml-tree-sitter-core/issues/8 ?

**Comment by aryx** [comment]
Not sure I understand the alternative solutions. Maybe a concrete example of what the CST look like and what other information the parser would return would help. In any case, the only thing we need semgrep-side is at least the information that somes parts of the CST are actually wrong because they
are missing nodes, instead of silently believing we parsed 100% correctly a file.

**Comment by mjambon** [comment]
> Not sure I understand the alternative solutions. Maybe a concrete example of what the CST look like and what other information the parser would return would help.

I went with the alternative 2, which is the simplest and most natural. The CST is unchanged. Bogus MISSING nodes are left in place.

> In any case, the only thing we need semgrep-side is at least the information that somes parts of the CST are actually wrong because they are missing nodes, instead of silently believing we parsed 100% correctly a file.

The list of errors is now the union of error nodes and missing nodes instead of just the error nodes. Their location correctly corresponds to the site of the error but since its length is zero, it may not be rendered well in some contexts (zero-length highlighted text is invisible; zero-length snippets extracted without context are an empty string).

**Comment by mjambon** [comment]
> Not sure I understand the alternative solutions. Maybe a concrete example of what the CST look like and what other information the parser would return would help.

I went with the alternative 2, which is the simplest and most natural. The CST is unchanged. Bogus MISSING nodes are left in place.

> In any case, the only thing we need semgrep-side is at least the information that somes parts of the CST are actually wrong because they are missing nodes, instead of silently believing we parsed 100% correctly a file.

The list of errors is now the union of error nodes and missing nodes instead of just the error nodes. Their location correctly corresponds to the site of the error but since its length is zero, it may not be rendered well in some contexts (zero-length highlighted text is invisible; zero-length snippets extracted without context are an empty string).

**Comment by mjambon** [comment]
Test grammar:
```js
module.exports = grammar({
  name: "report_missing",
  rules: {
    program: $ => repeat($.tuple),

    tuple: $ => choice(
      seq(
        'a', $.terminator
      ),
      seq(
        'b', ';'
      ),
      seq(
        'c', /;/
      ),
    ),
    terminator: $ => /[;?.]/
  }
});
```

This is the tree-sitter output (not great but shows zero-width locations) for a missing constant (`;`):
```
$ tree-sitter-prod parse <(echo 'a; b')
(program [0, 0] - [1, 0]
  (tuple [0, 0] - [0, 2]
    (terminator [0, 1] - [0, 2]))
  (tuple [0, 3] - [0, 4]))
/dev/fd/63	0 ms	(MISSING ";" [0, 4] - [0, 4])
```
and for a missing named rule (`$.terminator`):
```
$ tree-sitter-prod parse <(echo 'b; a')
(program [0, 0] - [1, 0]
  (tuple [0, 0] - [0, 2])
  (tuple [0, 3] - [0, 4]
    (terminator [0, 4] - [0, 4])))
/dev/fd/63	0 ms	(MISSING terminator [0, 4] - [0, 4])
```

In the case of an unnamed, non-constant terminal (patterns such as `/[;.]/`), tree-sitter doesn't produce a MISSING node at all even though a node is really missing. It's an old issue with unnamed nodes that ocaml-tree-sitter avoids by assigning names to patterns (with `alias()`):
```
(program [0, 0] - [1, 0]
  (tuple [0, 0] - [0, 2]
    (terminator [0, 1] - [0, 2]))
  (tuple [0, 3] - [0, 4]))
```

The output of the ocaml-tree-sitter test parser shows a missing node with a generated name:

```
$ ./parse <(echo 'a; c')
CST obtained from the tree-sitter parser:
program:
| tuple:
|   "a"
|   terminator
| tuple:
|   "c"
|   pat_9eecb7d
---
Recovered typed CST:
[
  A_term (
    (
      "a"
      ";"
    )
  )
  C_pat_9eecb7d (
    (
      "c"
      ""
    )
  )
]
total lines: 2
error lines: 1
error count: 1
success: 50.00%
Error: File /dev/fd/63, line 1, characters 4-4:
a; c


parent: tuple, left sibling: "c", no children
node type: pat_9eecb7d
children: [
]
Missing element in input code: pat_9eecb7d
```

That is because the rewritten grammar has an alias construct giving a name to the pattern:
```js
$ cat grammar-rev.js
// JavaScript grammar recovered from JSON by 'ocaml-tree-sitter to-js'
module.exports = grammar({
  name: "report_missing",
  externals: $ => [
  ],
  conflicts: $ => [
  ],
  inline: $ => [
  ],
  precedences: $ => [
  ],
  supertypes: $ => [
  ],
  extras: $ => [
    /\s/
  ],
  rules: {
    program: $ => repeat(
      $.tuple
    ),
    tuple: $ => choice(
      seq(
        "a",
        $.terminator
      ),
      seq(
        "b",
        ";"
      ),
      seq(
        "c",
        alias(
          /;/,
          $.pat_9eecb7d
        )
      )
    ),
    terminator: $ => /[;?.]/,
    pat_9eecb7d: $ => /;/
  }
});
```

File: src/bindings/lib/Tree_sitter_output.atd [hunk 1]
```diff
@@ -32,6 +32,10 @@ type node = {
      tagged appropriately to avoid confusion. *)
   ~kind <ocaml default="Literal \"\"">: node_kind;
 
+  (* Missing nodes are suggested by tree-sitter. They don't exist in the
+     input. *)
+  ~is_missing: bool;
+
```
Comment (mjambon) [inline src/bindings/lib/Tree_sitter_output.atd:38]: The addition of a new field should not break older code unless it tries to pattern-match records exactly. I would be a problem only for pre-generated code (`semgrep-`* repos) but it's not the case.

**Review by aryx — APPROVED** [review]
great work
Our parsing stats may go down after this, but that's ok. Better to be correct.
I also wonder whether the position of those missing nodes in the CST will be problematic
for matching/autofix. They are a bit like fake tokens.
cc @nmote

**Pull request merged**